Guard ticket specific TLS 1.3 function with macro

inorick opened the pull request:
Guard `ssl_tls13_write_new_session_ticket_coordinate` with `MBEDTLS_SSL_SESSION_TICKETS` macro.

Enabling `MBEDTLS_SSL_PROTO_TLS1_3` and `MBEDTLS_USE_PSA_CRYPTO` while disabling `MBEDTLS_SSL_SESSION_TICKETS` causes compilation to fail. This commit fixes one of the causes.

This partially fixes #6395.

Signed-off-by: Norbert Fabritius <norbert.fabritius@esrlabs.com>

## PR checklist

- [x] **changelog** provided
- [x] **3.6 backport** required, #9014
- [x] **2.28 backport** not required - no TLS 1.3 in 2.28
- [x] **tests** provided

inorick commented:
The following errror remains:
```
ssl_tls13_server.c:562:25: error: ‘session’ undeclared (first use in this function)
  562 |             &psk_type, &session);
```

However, I am not sure how to fix this properly. The function `ssl_tls13_parse_pre_shared_key_ext` conditionally defines the variable `sessions`:
```
#if defined(MBEDTLS_SSL_SESSION_TICKETS)
        mbedtls_ssl_session session;
        mbedtls_ssl_session_init(&session);
#endif
```

but later uses it unconditionally:
```
ret = ssl_tls13_offered_psks_check_identity_match(
            ssl, identity, identity_len, obfuscated_ticket_age,
            &psk_type, &session);
```

ronald-cron-arm commented on library/ssl_tls13_server.c:
```diff
@@ -2586,6 +2586,7 @@ static int ssl_tls13_handshake_wrapup(mbedtls_ssl_context *ssl)
     return 0;
 }
 
+#if defined(MBEDTLS_SSL_SESSION_TICKETS)
```
This looks good to me.

ronald-cron-arm commented:
> However, I am not sure how to fix this properly. The function `ssl_tls13_parse_pre_shared_key_ext` conditionally [defines](https://github.com/Mbed-TLS/mbedtls/blob/development/library/ssl_tls13_server.c#L537) the variable `sessions`:
> 
> ```
> #if defined(MBEDTLS_SSL_SESSION_TICKETS)
>         mbedtls_ssl_session session;
>         mbedtls_ssl_session_init(&session);
> #endif
> ```

I'd say to we can afford the extra unnecessary usage of the stack if the support for tickets is not enabled thus I think I would just remove  "#if defined(MBEDTLS_SSL_SESSION_TICKETS)".

paul-elliott-arm commented:
@inorick - if you want to add the other fix to this PR, then we can get them both in at the same time, potentially.

inorick commented:
@ronald-cron-arm I removed the suggested line. I also fixed all compiler warnings that I got when compiling with `MBEDTLS_SSL_PROTO_TLS1_3` and `MBEDTLS_USE_PSA_CRYPTO` but without `MBEDTLS_SSL_SESSION_TICKETS`.

inorick commented:
I rebased the PR onto the current `development` branch. Is there anything missing to push this forward?

yuhaoth reviewed (changes requested):
Would you like add test components in `tests/scripts/all.sh` ?

inorick commented:
@yuhaoth Sure, I added `component_test_tls13_no_session_ticket` to `all.sh`. I tried to copy the structure of existing tests.

yuhaoth commented:
> @yuhaoth Sure, I added `component_test_tls13_no_session_ticket` to `all.sh`. I tried to copy the structure of existing tests.

CI found some failure in `component_test_tls13_no_session_ticket` . It is due to exists issues. 
Would you like change `ssl-opt.sh` to fix it? It can be fixed with `requires_config_enabled MBEDTLS_SSL_SESSION_TICKETS`. 

Bellow is an example.
```
requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3
requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED
requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
requires_config_enabled MBEDTLS_SSL_SESSION_TICKETS
run_test    "Non-blocking I/O: TLS 1.3 + ticket + client auth + resume" \
            "$P_SRV nbio=2 tickets=1 auth_mode=required" \
            "$P_CLI nbio=2 tickets=1 reconnect=1" \
            0 \
            -S "mbedtls_ssl_handshake returned" \
            -C "mbedtls_ssl_handshake returned" \
            -c "Read from server: .* bytes read"
```

yuhaoth commented:
And I am try to fix the CI failure at https://github.com/Mbed-TLS/mbedtls/pull/7356 . Please take a look.

inorick commented:
`HEAD~7` points to `development` as this PR adds exactly 7 commits. Do you want me to rebase on the current `development` branch instead?

yuhaoth commented:
> `HEAD~7` points to `development` as this PR adds exactly 7 commits. Do you want me to rebase on the current `development` branch instead?

Yes

yuhaoth commented:
Not only one test case fail, other fails are same issue. 

I put my fix in https://github.com/Mbed-TLS/mbedtls/pull/7356/commits/541e7c5c0e83608c19e8e4603be21333fb75c173 . Please take a look.

ronald-cron-arm commented:
> I cherry picked your commit but it still seems to fail.

One ssl-opt.sh test case is still failing: "DTLS reordering: Buffer out-of-order CCS message on client". It seems it has a dependency on MBEDTLS_SSL_SESSION_TICKETS through the test argument "$P_PXY delay_srv=NewSessionTicket".

ronald-cron-arm commented on tests/scripts/all.sh:
```diff
@@ -3878,6 +3878,19 @@ component_test_tls13_no_compatibility_mode () {
+component_test_tls13_no_session_ticket() {
+    msg "build: default config with MBEDTLS_SSL_PROTO_TLS1_3, without session tickets"
+    scripts/config.py unset MBEDTLS_SSL_SESSION_TICKETS
+    scripts/config.py set MBEDTLS_SSL_PROTO_TLS1_3
```
```suggestion
    scripts/config.py set MBEDTLS_SSL_PROTO_TLS1_3
    scripts/config.py set MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
```
otherwise most of the compatibility test cases are skipped. That's probably something we should improve but for the time being that's how it is.

inorick commented:
I see
```
DTLS reordering: Buffer out-of-order CCS message on client ............. FAIL
```
:shrug:

ronald-cron-arm commented:
> I see
> 
> ```
> DTLS reordering: Buffer out-of-order CCS message on client ............. FAIL
> ```
> 
> shrug

Please see https://github.com/Mbed-TLS/mbedtls/pull/6955#issuecomment-1508456283. I think this test depends on MBEDTLS_SSL_SESSION_TICKETS as well.

inorick commented:
Looks like it works now. Anything else?

ronald-cron-arm commented:
> Looks like it works now. Anything else?

Regarding CI, the last commit is missing the "Signed-off-by" line.

yuhaoth reviewed (approved):
LGTM

inorick commented:
@ronald-cron-arm anything else to do from your perspective?

KloolK commented:
It would be great if this could be merged soon, since this issue make updating quite complicated for us.

@ronald-cron-arm Is there anything left that needs to be done here? Could you retrigger the CI, please? Locally, the changes still apply cleanly for me. However, I did not run all the tests.

ronald-cron-arm commented:
Hi, the PR seems in a good shape, it is just it got out of my radar unfortunately. Sorry about that. This PR is in the EPIC (TLS 1.3 misc for 3.6 LTS) that we have planned to work on this quarter. Thus this should be merged quite soon I'd say.

tom-cosgrove-arm commented:
@inorick Apologies for taking so long to look at this. Would you mind fixing the conflicts so that we can re-review with a goal to merging this soon? Thanks

inorick commented:
@tom-cosgrove-arm I tried rebasing this PR and got it to compile. However, I saw 111 test cases failing so this might take a moment.

ronald-cron-arm commented:
Hi, I have been looking at the rebase of this PR and making it pass the CI. @inorick would it be okay on your side if I push what I have to move this work forward (and sorry again for not having been able to progress faster on this PR).

inorick commented:
> @inorick would it be okay on your side if I push what I have to move this work forward (and sorry again for not having been able to progress faster on this PR).

Please feel free to go ahead. I am glad if we can this merged. Thank you.

ronald-cron-arm commented:
> > @inorick would it be okay on your side if I push what I have to move this work forward (and sorry again for not having been able to progress faster on this PR).
> 
> Please feel free to go ahead. I am glad if we can this merged. Thank you.

Great, thanks.

ronald-cron-arm commented:
@tom-cosgrove-arm for your information, CI is green, this is ready for review.

tom-cosgrove-arm reviewed (dismissed):
LGTM, one minor comment, happy if you want to keep the code the way it is

mpg commented on tests/scripts/all.sh:
```diff
@@ -6050,6 +6050,17 @@ component_test_tls13_no_compatibility_mode () {
     tests/ssl-opt.sh
 }
 
+component_test_default_minus_session_tickets() {
```
My instinct would be to go for full minus session tickets, rather than default. Though I'm not sure that makes much of a difference now, so I won't insist on it if you disagree.

ronald-cron-arm replied:
No I do not have any objections to that, changed. The main difference is probably USE_PSA_CRYPTO being enabled or not.

mpg reviewed (changes requested):
Looking pretty good to me, just one question and a likely copy-paste mistake.

ronald-cron-arm commented on tests/suites/test_suite_ssl.data:
```diff
@@ -930,35 +930,35 @@ depends_on:MBEDTLS_SSL_PROTO_TLS1_2
 
 TLS 1.3: CLI: session serialization: Wrong minor version
-depends_on:MBEDTLS_SSL_PROTO_TLS1_3:MBEDTLS_SSL_CLI_C
+depends_on:MBEDTLS_SSL_PROTO_TLS1_3:MBEDTLS_SSL_CLI_C:MBEDTLS_SSL_SESSION_TICKETS
```
Because in ssl_tls.c, if MBEDTLS_SSL_SESSION_TICKETS is not defined, `ssl_tls13_session_save/load()` are just returning `MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE`. The reason is that in TLS 1.3 session resumption is supported only through tickets, not though session id.

mpg replied:
Ok, I see. Seems like something that the documentation of `mbedtls_ssl_session_save/load()` should mention. I understand the reasoning but I can also see how that can surprise users: say I'm using 1.2 in a config without tickets (using cache-based resumption), I'm calling those functions unconditionally client-side and they work; now I enable 1.3 (or inherit it from the default config), so 1.3 is negotiated and boom `mbedtls_ssl_session_save()` starts returning an error. So, I'd double-check this function's documentation and when finding no hint there I'd probably not be super happy.

ronald-cron-arm replied:
Thanks for raising this. Looking at this, it seems to me that it would also make sense to return in error when `mbedtls_ssl_get_session()` is called on an TLS 1.3 session and MBEDTLS_SSL_SESSION_TICKETS is disabled (not the case currently). What do you think?

mpg replied:
I don't know. On one hand, I see your point, and I can't think of a reason why people would want to use `get_session()` with TLS 1.3 without tickets - you can't resume anyway. On the other hand, part of me is worried that I have less imagination than users and they might still want to do something with it.

So, for 3.6 I'm inclined to only document what wasn't working anyway (save/load) but not artificially break what was working (get) even if we think it's not useful anyway.

For 4.0 as discussed previously I'd like to refactor this whole API and combine get+save and load+set into a single function each, so that users don't have to manipulate `ssl_session` directly.

Wdyt?

ronald-cron-arm replied:
That sounds sensible to me, thanks. I will update the documentation of save/load.

mpg reviewed (commented):
Thanks for addressing my feedback. At cf47a15e960d279d9e18d6728a9961942843a5d7 looking good to me except for this conversation pending resolution: https://github.com/Mbed-TLS/mbedtls/pull/6955#discussion_r1547644694

ronald-cron-arm commented:
> Thanks for addressing my feedback. At [cf47a15](https://github.com/Mbed-TLS/mbedtls/commit/cf47a15e960d279d9e18d6728a9961942843a5d7) looking good to me except for this conversation pending resolution: [#6955 (comment)](https://github.com/Mbed-TLS/mbedtls/pull/6955#discussion_r1547644694)

This turns out to be a bit of a rabbit hole. I've fixed/changed the documentation of mbedtls_ssl_set/get_session() and mbedtls_ssl_session_save(). I did not feel the need to change the session_load() one eventually as there is not such a thing as a serialized TLS 1.3 session when session tickets are disabled.

tom-cosgrove-arm commented on include/mbedtls/ssl.h:
```diff
@@ -3309,8 +3309,16 @@ int mbedtls_ssl_session_load(mbedtls_ssl_session *session,
  *                 with \p buf set to \c NULL and \p buf_len to \c 0.
  *
+ * \note           For TLS 1.3 sessions, this feature is supported only if the
+ *                 MBEDTLS_SSL_SESSION_TICKETS configuration option is enabled,
```
Should this be `MBEDTLS_SSL_SESSION_TICKETS` to create a Doxygen link? (same on line 3320)

ronald-cron-arm commented:
CI passed but a spurious semi-colon in the change log. Ready for further review.

tom-cosgrove-arm reviewed (approved):
LGTM

mpg reviewed (approved):
LGTM, thanks! And sorry for sending you down a rabbit hole ^^

ronald-cron-arm commented:
> LGTM, thanks! And sorry for sending you down a rabbit hole ^^

That things we needed to do and I am rather glad that your comments pointed me to the rabbit hole. Thanks.
